File: level.1_argorithm/crane_doll/crane_doll.java
import java.util.ArrayList; 

class crane_doll {
    static public int solution(int[][] board, int[] moves) {
        int answer = 0;
        ArrayList<Integer> before = new ArrayList<Integer>();
        for(int move : moves)
        {
          for(int i=0; i<board.length; i++)
          {
            if(board[i][move] !=0)
            {
              before.add(board[i][move]);
              if(before.size() >1)
              {
                if(before.get(before.size()) == before.get(before.size() - 1)){
                  before.remove(before.size());
                  before.remove(before.size()-1);
                  answer +=2;
                }
              }
            }
          }
        }
        return answer;
    }

    public static void main(String args[])
    {
      int[][] board = {{0,0,0,0,0},{0,0,1,0,3},{0,2,5,0,1},{4,2,4,4,2},{3,5,1,3,1}};
      int[] moves = {1,5,3,5,1,2,1,4};
      int answer = 0;
      ArrayList<Integer> before = new ArrayList<Integer>();
      for(int move : moves)
      {
        for(int i=0; i<board.length; i++)
        {
          if(board[i][move] !=0)
          {
            before.add(board[i][move]);
            if(before.size() >1)
            {
              if(before.get(before.size()) == before.get(before.size() - 1)){
                before.remove(before.size());
                before.remove(before.size()-1);
                answer +=2;
              }
            }
          }
        }
      }
      System.out.println(answer);
    }
}

File: level.1_argorithm/crane_doll/crane_doll.py
#https://programmers.co.kr/learn/courses/30/lessons/64061?language=python
# 인형은 1~5의 종류가 있으며0은 없는 칸이다. 

def solution(board, moves):
  answer = 0
  # basket에 -10을 넣어준 이유는 -1인덱스 접근시 out of list 에러를 방지하기 위해서이다.
  basket =[-10]
  for move in moves:
    for i in range(len(board[0])):
      print('crane: '+str(board[i][move-1]), end=' ')
      print('basket: '+str(basket),end=' ')
      print('move: '+str(+move),end=' ')
      print('answer: '+str(answer))
      if board[i][move-1] != 0:
        if basket[-1] != board[i][move-1]:
          basket.append(board[i][move-1])
          board[i][move-1] = 0
          break
        else:
          board[i][move-1] = 0
          basket.pop()
          answer+=2
          break
  return answer
  

print(solution([[0,0,0,0,0],[0,0,1,0,3],[0,2,5,0,1],[4,2,4,4,2],[3,5,1,3,1]], [1,5,3,5,1,2,1,4]))
